content/posts/pandastutor.md

```diff
@@ -0,0 +1,323 @@
+they didn't have wheels available on PyPI (e.g.,
+`micropip.install('docopt')` doesn't work). Our solution was to bundle
+(i.e., inline) all dependencies into a self-contained Pandas
+Tutor wheel so that our web frontend code needs only one call to
+`micropip` instead of almost a dozen separate calls:
+
+```js
+await pyodide.runPythonAsync(`
+import micropip
+await micropip.install('https://pandastutor.com/build/pandastutor-1.0-py3-none-any.whl')
+`);
+```
+
+This approach also has the nice side-effect of pinning all the
+dependency versions for better API stability. The only dependency we
+didn't bundle was pandas because that requires WASM-compiled code. But
+fortunately Pyodide comes with pandas
+[built-in](https://pyodide.org/en/stable/usage/packages-in-pyodide.html)!
+
+
+### Step 2: create a new Pyodide entry point
+
+The only modification we had to make to the Pandas Tutor codebase was
+adding a new `pyodide_main.py` file as the entry point for Pyodide:
+
+```py
+# simplified contents of pandas_tutor/pyodide_main.py
+from .main import run_user_code
+import js # to interface with pyodide
+def run_code_from_js():
+    return run_user_code(js.globalUserCode)
+```
+
+This entry point uses the `js` module to access JavaScript global
+variables. On the JavaScript end of our web frontend code, we created a
+corresponding function that takes the user's code to run (`userCode`)
+and calls the `run_code_from_js()` Python function:
+
+```js
+function runPyodidePandastutor(userCode) {
+  window.globalUserCode = userCode; // pass function argument as a global variable
+  const result = pyodide.runPython('pandas_tutor.pyodide_main.run_code_from_js()');
+  window.globalUserCode = undefined; // null this out afterward to prevent leakage
+  return result;
+}
+```
+
+We couldn't figure out a more elegant way to pass arguments from
+JavaScript to Python functions, so we used a `globalUserCode` variable
+as a hack that seems to work for now.
+
+
+### Step 3: import the Pandas Tutor module in JavaScript
+
+Our final setup step was adding a line of code to our frontend
+JavaScript to import the Pandas Tutor module:
+
+```js
+pyodide.runPython('import pandas_tutor.pyodide_main');
+```
+
+This is a synchronous call that imports `pandas_tutor` and all
+dependencies when the web app first loads. This line takes 3-5 seconds
+to finish since dependencies (including pandas) must be loaded.
+Unfortunately, browser caching can't speed things up here since even if
+all the files are cached locally, Pyodide still needs to run the import
+and initialization code at page load time.
+
+One feature addition that might speed up this process is if Pyodide
+could somehow *serialize a binary memory snapshot* taken right after
+everything gets initialized. Then the user's browser can directly load
+that snapshot on the first page visit and cache it for future visits.
+(But maybe that snapshot would be too big to download efficiently, so
+nevermind!)
```

Comment:
The bottleneck for loading pandas is mostly WASM compilation by the browser. It should be cached in Chrome on page reload. But yeah there are certainly lots of ideas to explore in that area https://github.com/pyodide/pyodide/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+size+label%3A%22size+%26+load+time+optimization%22